src/main/java/de/tum/in/www1/artemis/service/connectors/pyris/PyrisPipelineService.java

```diff
@@ -39,58 +55,103 @@ public class PyrisPipelineService {
+                pyrisConnectorService.executePipeline(pipelineName, variant, executionDTO);
             }
             catch (PyrisConnectorException | IrisException e) {
-                log.error("Failed to execute tutor chat pipeline", e);
+                log.error("Failed to execute " + pipelineName + " pipeline", e);
                 var executingPipelineStageFailed = new PyrisStageDTO("Executing pipeline", 30, PyrisStageStateDTO.ERROR, "An internal error occurred");
                 irisChatWebsocketService.sendStatusUpdate(session, List.of(preparingRequestStageDone, executingPipelineStageFailed));
             }
         }
         catch (Exception e) {
-            log.error("Failed to prepare tutor chat pipeline execution", e);
+            log.error("Failed to prepare " + pipelineName + " pipeline execution", e);
             var preparingRequestStageFailed = new PyrisStageDTO("Preparing request", 10, PyrisStageStateDTO.ERROR, "An internal error occurred");
             irisChatWebsocketService.sendStatusUpdate(session, List.of(preparingRequestStageFailed, executingPipelineStageNotStarted));
         }
+    }
+
+    public void executeExerciseChatPipeline(String variant, Optional<ProgrammingSubmission> latestSubmission, ProgrammingExercise exercise, IrisExerciseChatSession session) {
+        executeChatPipeline(variant, session, "exercise-chat",
+                base -> new PyrisExerciseChatPipelineExecutionDTO(base, latestSubmission.map(pyrisDTOService::toPyrisSubmissionDTO).orElse(null),
+                        pyrisDTOService.toPyrisProgrammingExerciseDTO(exercise), new PyrisCourseDTO(exercise.getCourseViaExerciseGroupOrCourseMember())),
+                () -> pyrisJobService.addExerciseChatJob(exercise.getCourseViaExerciseGroupOrCourseMember().getId(), exercise.getId(), session.getId()));
+    }
+
+    public void executeCourseChatPipeline(String variant, IrisCourseChatSession session) {
+        var courseId = session.getCourse().getId();
+        var studentId = session.getUser().getId();
+        executeChatPipeline(variant, session, "course-chat", base -> {
+            var fullCourse = loadCourseWithParticipationOfStudent(courseId, studentId);
+            var metrics = metricsService.getStudentCourseMetrics(session.getUser().getId(), courseId);
+            return new PyrisCourseChatPipelineExecutionDTO(base, PyrisExtendedCourseDTO.of(fullCourse), metrics);
+        }, () -> pyrisJobService.addCourseChatJob(courseId, session.getId()));
+    }
+
+    /**
+     * Load the course with the participation of the student and set the participations on the exercises.
+     * <p>
+     * Spring Boot 3 does not support conditional left joins, so we have to load the participations separately.
+     *
+     * @param courseId  the id of the course
+     * @param studentId the id of the student
+     */
+    private Course loadCourseWithParticipationOfStudent(Long courseId, Long studentId) {
```

Comment:
why boxed types if the method doesn't make sense to be called without student and course ids?
```suggestion
    private Course loadCourseWithParticipationOfStudent(long courseId, long studentId) {
```

Reply:
No idea, lets change it